File: leetcode/02-Add-Two-Numbers/python/main.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def convertArrayToListNode(nums):
  if len(nums) == 1:
      return ListNode(nums[0])
  return ListNode(nums[0], convertArrayToListNode(nums[1:]))

def convertListNodeToArray(nodes):
    result = []
    node = nodes
    while node:
        result.append(node.val)
        if node.next:
            node = node.next
    return result

class Solution:
    def addTwoNumbers(self, l1, l2):
        result = []
        currentNodeL1 = l1
        currentNodeL2 = l2
        carry = 0

        while currentNodeL1 or currentNodeL2:
            if not currentNodeL1 is None and not currentNodeL2 is None:
              sum = currentNodeL1.val + currentNodeL2.val + carry
              currentNodeL1 = currentNodeL1.next
              currentNodeL2 = currentNodeL2.next
            elif currentNodeL2 is None:
              sum = currentNodeL1.val + carry
              currentNodeL1 = currentNodeL1.next
            elif currentNodeL1 is None:
              sum = currentNodeL2.val + carry
              currentNodeL2 = currentNodeL2.next
                
            carry = 1 if sum >= 10 else 0
            low = sum % 10
            result.append(low)
        
        if carry == 1:
           result.append(1)

        return convertArrayToListNode(result)

File: leetcode/02-Add-Two-Numbers/go/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func convertToListNodeFromList(l []int) *ListNode {
	if len(l) == 1 {
		node := &ListNode{l[0], nil}
		return node
	}
	nodes := &ListNode{l[0], convertToListNodeFromList(l[1:])}
	return nodes
}

func convertToListFromNodeList(nodes *ListNode) []int {
	resultList := []int{}
	currentNode := nodes

	for currentNode != nil {
		resultList = append(resultList, currentNode.Val)
		currentNode = currentNode.Next
	}

	return resultList
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	currentL1Node := l1
	currentL2Node := l2
	carry := 0
	tmpSum := 0
	resultList := []int{}

	for {
		if currentL1Node != nil && currentL2Node != nil {
			tmpSum = currentL1Node.Val + currentL2Node.Val + carry
			currentL1Node = currentL1Node.Next
			currentL2Node = currentL2Node.Next
		} else if currentL1Node != nil {
			tmpSum = currentL1Node.Val + carry
			currentL1Node = currentL1Node.Next
		} else if currentL2Node != nil {
			tmpSum = currentL2Node.Val + carry
			currentL2Node = currentL2Node.Next
		} else {
			break
		}

		if tmpSum >= 10 {
			carry = 1
		} else {
			carry = 0
		}
		resultList = append(resultList, tmpSum%10)
	}

	if carry > 0 {
		resultList = append(resultList, 1)
	}

	return convertToListNodeFromList(resultList)
}
